server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -72,6 +76,70 @@ public void execute(SearchContext context, int[] docIdsToLoad, RankDocShardInfo
         execute(context, docIdsToLoad, rankDocs, null);
     }
 
+    public void execute(
+        SearchContext context,
+        int[] docIdsToLoad,
+        RankDocShardInfo rankDocs,
+        @Nullable IntConsumer memoryChecker,
+        @Nullable FetchPhaseResponseChunk.Writer writer
+    ) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("{}", new SearchContextSourcePrinter(context));
+        }
+
+        if (context.isCancelled()) {
+            throw new TaskCancelledException("cancelled");
+        }
+
+        final Profiler profiler = context.getProfilers() == null
+            || (context.request().source() != null && context.request().source().rankBuilder() != null)
+                ? Profiler.NOOP
+                : Profilers.startProfilingFetchPhase();
+
+        SearchHits hits = null;
+        try {
+            // Collect all pending chunk futures
+            final int maxInFlightChunks = 3; // TODO make configurable
+            final ArrayDeque<PlainActionFuture<Void>> pendingChunks = new ArrayDeque<>();
+            final AtomicReference<Throwable> sendFailure = new AtomicReference<>();
+            hits = buildSearchHits(
+                context,
+                docIdsToLoad,
+                profiler,
+                rankDocs,
+                memoryChecker,
+                writer,
+                pendingChunks,
+                maxInFlightChunks,
+                sendFailure
+            );
+
+            // Wait for all chunks to be ACKed before setting final result
+            if (writer != null && pendingChunks.isEmpty() == false) {
+                try {
+                    // Wait for all pending chunks sequentially
+                    for (PlainActionFuture<Void> future : pendingChunks) {
+                        future.actionGet();
```

Comment:
I'm suspicious about the need to block the whole search thread here. I would expect this can be done properly asynchronously, with the subsequent calls to `context.fetchResult().shardResult()` etc either done on the thread that receives the final response (if it's fast) or dispatched back to a search coordination thread by that response (if it's not). We should be using this thread for other things in the meantime.

I would think that a `org.elasticsearch.action.support.RefCountingListener` is what you want here instead of all these futures.